aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -410,6 +427,177 @@ impl AptosVM {
+            bcs::to_bytes::<Vec<u8>>(&vec![]).unwrap()
+        };
+        // Failures here will be propagated back.
+        let payload_bytes: Vec<Vec<u8>> = session
+            .execute_function_bypass_visibility(
+                &MULTISIG_ACCOUNT_MODULE,
+                GET_NEXT_TRANSACTION_PAYLOAD,
+                vec![],
+                serialize_values(&vec![
+                    MoveValue::Address(txn_payload.multisig_address),
+                    MoveValue::vector_u8(provided_payload.clone()),
+                ]),
+                gas_meter,
+            )?
+            .return_values
+            .into_iter()
+            .map(|(bytes, _ty)| bytes)
+            .collect::<Vec<_>>();
+        // We have to deserialize once to get the actual return type of the function, which is
+        // vec<u8>.
+        let payload: Vec<u8> = bcs::from_bytes(payload_bytes.first().unwrap()).map_err(|_| {
+            PartialVMError::new(StatusCode::UNREACHABLE).finish(Location::Undefined)
+        })?;
+        // Deserialize a second time to convert bytes into the actual multisig payload type.
+        let payload: MultisigTransactionPayload = bcs::from_bytes(&payload).map_err(|_| {
+            PartialVMError::new(StatusCode::UNREACHABLE).finish(Location::Undefined)
+        })?;
+
+        // Step 2: Execute the target payload. Transaction failure here is tolerated. In case of any
+        // failures, we'll discard the session and start a new one. This ensures that any data
+        // changes are not persisted.
+        // The multisig transaction would still be considered executed even if execution fails.
+        let execution_result = self.execute_multisig_entry_function(
+            &mut session,
+            gas_meter,
+            txn_payload.multisig_address,
+            &payload,
+        );
+
+        // Step 3: Call post transaction cleanup function in multisig account module with the result
+        // from Step 2.
+        let mut cleanup_args = serialize_values(&vec![
+            MoveValue::Address(txn_data.sender),
+            MoveValue::Address(txn_payload.multisig_address),
+            MoveValue::vector_u8(provided_payload),
+        ]);
+        let session_output = if let Err(execution_error) = execution_result {
+            let execution_error = Self::convert_to_execution_error(execution_error)?;
+            cleanup_args.push(bcs::to_bytes(&execution_error).unwrap());
+
+            // Reset the session to skip any state changes from the failed execution.
+            // The gas meter is not reset, so sender would still be charged for the failed execution
+            let mut new_session = self.0.new_session(storage, SessionId::txn_meta(txn_data));
+            new_session.execute_function_bypass_visibility(
+                &MULTISIG_ACCOUNT_MODULE,
+                FAILED_TRANSACTION_EXECUTION_CLEANUP,
+                vec![],
+                cleanup_args,
+                gas_meter,
```

Comment:
Is it possible to have out of gas in step 3?

Reply:
It's possible. In the current implementation, I think that'd lead to the entire tx failing (and thus retriable with higher gas limit). I can add a test that demonstrate this and we can discuss whether this being retriable is a desired behavior.